shared/js/passcode_helper.js

```diff
@@ -0,0 +1,147 @@
+/* exported PasscodeHelper */
+/* globals crypto, TextEncoder */
+(function(exports) {
+  'use strict';
+
+  const SET_DIGEST_VALUE = 'lockscreen.passcode-lock.digest.value';
+  const SET_DIGEST_SALT = 'lockscreen.passcode-lock.digest.salt';
+  const SET_DIGEST_ITERATIONS = 'lockscreen.passcode-lock.digest.iterations';
+  const SET_DIGEST_ALGORITHM = 'lockscreen.passcode-lock.digest.algorithm';
+  const DEFAULT_ALGORITHM = 'SHA-1';//XXX Update to SHA-256 after bug 554827
+  const DEFAULT_ITERATIONS = 1000;
+
+  var PasscodeHelper = {
+    /*
+     * PasscodeHelper.setPassccode(string) -> Promise => digest (or false)
+     * PasscodeHelper.checkPasscode(string) -> Promise resolves to a boolean
+     * */
+
+    _encode: function(str) {
+      return new TextEncoder('utf-8').encode(str);
+    },
+    _toTypedArray: function(obj) {
+      // SettingsAPI doesnt like arrays and gives us an Object { 1: .., 2: .. }
+      var a = [];
+      for (var key in obj) {
+        a.push(obj[key]);
+      }
+      return new Uint8Array(a);
+    },
+    _deriveBits: function(pwKey, salt, iterations, algorithm) {
+      var length = 256;
+      var params = {
+        name: 'PBKDF2',
+        hash: algorithm,
+        salt: salt,
+        iterations: iterations
+      };
+      return crypto.subtle.deriveBits(params, pwKey, length);
+    },
+    _makeDigest: function(pass, salt, iterations, algorithm) {
+      var bytes = this._encode(pass);
```

Comment:
Yeah, you're right. I was a bit lost when I realized that I needed to retain state. I'll rewrite this bit.